src/kbmod/search/stack_search.cpp

```diff
@@ -155,63 +155,81 @@ Trajectory StackSearch::search_linear_trajectory(short x, short y, float vx, flo
     return result;
 }
 
-void StackSearch::search(std::vector<Trajectory>& search_list, int min_observations) {
-    DebugTimer core_timer = DebugTimer("core search", rs_logger);
+void StackSearch::finish_search(){
+    psi_phi_array.clear_from_gpu();
+    gpu_search_list.move_to_cpu();
+}
 
-    DebugTimer psi_phi_timer = DebugTimer("creating psi/phi buffers", rs_logger);
+void StackSearch::prepare_search(std::vector<Trajectory>& search_list, int min_observations){
+    DebugTimer psi_phi_timer = DebugTimer("Creating psi/phi buffers", rs_logger);
     prepare_psi_phi();
     psi_phi_array.move_to_gpu();
     psi_phi_timer.stop();
 
-    // Allocate a vector for the results and move it onto the GPU.
-    int search_width = params.x_start_max - params.x_start_min;
-    int search_height = params.y_start_max - params.y_start_min;
-    int num_search_pixels = search_width * search_height;
-    int max_results = num_search_pixels * RESULTS_PER_PIXEL;
+    
+    int num_to_search = search_list.size();
+    if (debug_info) std::cout << "Preparing to search " << num_to_search << " trajectories... \n" << std::flush;
+    gpu_search_list.set_trajectories(search_list);
+    gpu_search_list.move_to_gpu();
+
+    params.min_observations = min_observations;
+}
+
+void StackSearch::search_all(std::vector<Trajectory>& search_list, int min_observations) {
+    prepare_search(search_list, min_observations);
+    search_batch();
+    finish_search();
+}
+
+void StackSearch::search_batch(){
+    if(!psi_phi_array.gpu_array_allocated()){
+        throw std::runtime_error("PsiPhiArray array not allocated on GPU. Did you forget to call prepare_search?");
+    }
+
+    DebugTimer core_timer = DebugTimer("Running batch search", rs_logger);
+    int max_results = extract_max_results();
+
     // staple C++
     std::stringstream logmsg;
     logmsg << "Searching X=[" << params.x_start_min << ", " << params.x_start_max << "] "
            << "Y=[" << params.y_start_min << ", " << params.y_start_max << "]\n"
            << "Allocating space for " << max_results << " results.";
     rs_logger->info(logmsg.str());
 
+
     results.resize(max_results);
     results.move_to_gpu();
 
-    // Allocate space for the search list and move that to the GPU.
-    int num_to_search = search_list.size();
-
-    logmsg.str("");
-    logmsg << search_list.size() << " trajectories...";
-    rs_logger->info(logmsg.str());
-
-    TrajectoryList gpu_search_list(search_list);
-    gpu_search_list.move_to_gpu();
-
-    // Set the minimum number of observations.
-    params.min_observations = min_observations;
-
-    // Do the actual search on the GPU.
-    DebugTimer search_timer = DebugTimer("search execution", rs_logger);
+        // Do the actual search on the GPU.
```

Comment:
Indentation of this comment appears off.

Reply:
Cleaned